### DISCUSSION TITLE
批次管理功能 Lote

### DISCUSSION THREAD

**Discussion by densen2014** [densen2014/docs#28] [想法]
## 开启设置

![image](https://github.com/densen2014/docs/assets/8428709/9e8da137-bf8d-4ec1-a7db-2ee4598a58e6)

## 管理批次

![image](https://github.com/densen2014/docs/assets/8428709/e02f2cc2-34ec-4e1b-8ff7-6e7963170c33)

## 销售点击[批次]列单元格选择批次

为了提高性能, 选择批次窗口默认不会被关闭

![image](https://github.com/densen2014/docs/assets/8428709/44613aac-4a8e-4a35-a1ee-33fabd59fd7f)

## 结账扣数量和编辑回退数量没做

等确认功能

## 查询销售历史功能没做

等确认功能

## 疑问

是否需要添加日期，最后使用日期字段

**Comment by DENSEN-BIN** [comment]
1. 快速录入需要直接填写批次导入商品，方便存货
2. 进货界面也需要和销售界面一样的选择批次功能（需要进老货的时候，可以在进货的界面直接操作）
![image](https://github.com/densen2014/docs/assets/69311377/288d4619-7a9c-495b-9431-f30977506a86)
3. 批次的库存调整，举例：0号商品的批次 123 库存显示100，实际只有90.
                                           0号商品的批次 123456库存显示150，实际是160（因为他们拿货物的时候，拿错了）
  需要在进货界面调整批次的数量。批次123 减少10，批次123456 增加10
![image](https://github.com/densen2014/docs/assets/69311377/82ad7180-c64d-47f3-9629-99d930791ed9)

**Reply by densen2014** [reply]
批次数量严格管理的话, 工程就大了. 比如销售数量大于选择的批次数量,怎么办

**Reply by DENSEN-BIN** [reply]
出现一个提醒，弹出选项框，再次选择其他批次的货物。

**Comment by DENSEN-BIN** [comment]
4. 批次栏 需要过期日期时间小到大排序。库存大于0.当库存数量小于等于0，排列到最后
![image](https://github.com/densen2014/docs/assets/69311377/fdb37efa-450f-486a-8320-48e8f10df20d)
5. 客户改成备注，备注商品的销售情况，方便下一个销售员查看
![image](https://github.com/densen2014/docs/assets/69311377/396160ff-25b1-4178-821a-8d18b442a581)
6. 需要一个界面查询每个批次的进价。
7. 批次界面需要显示一个到期日期的天数。批次是TTTT/01/10/2023, 今天日期是06/10/2023，天数栏显示 5，方便销售的时候快速选择快过期的商品进行销售
8. 删除批次的按键做一个设置可以开启或者关闭。或者隐藏。
9. 后期A4打印需要显示批次信息

**Reply by densen2014** [reply]
9. 后期A4打印需要显示批次信息
    **要配合报表格式才能打印Lote, 参考以前的issue**

**Comment by NeverCarryC** [comment]
## 问题
销售点击[批次]列单元格选择批次 的文本提出问题：
1. 销售界面，收银员是每次都得手动选择批次，还是默认选择某个批次的商品，扣除对应库存(哪里可以进行测试)
## 建议
 要不要遵循 先进先出（FIFO）原则，销售时默认先扣除最早到期的批次货物（批次越早，越快过期），然后让摆货人员摆放商品的时候，把快过期的放在前面

## 担忧
如果库存管理增加批次功能，相当于从底层重构进，销，存模块，工作量很大

**Reply by densen2014** [reply]
FIFO 还涉及数量,还得考虑数量不够新生成一行选择其他批次, 一步步来. 先做提醒数量不够,强制选择批次

**Reply by densen2014** [reply]
批次一直是迫在眉睫的事情, 但是输入单据人员工作量是巨大的.

**Reply by densen2014** [reply]
默认选择某个批次的商品现阶段调整为按照排序规则显示在第一条

**Comment by NeverCarryC** [comment]


**Reply by densen2014** [reply]
可以, 你直接更新到测试版

**Comment by DENSEN-BIN** [comment]
建议很好，就是有网这个方面想

10. 先进先出（FIFO）原则，销售时默认先扣除最早到期的批次货物（批次越早，越快过期）
11. 首选默认最早的批次货物，当批次数量小于销售数量，出现提醒，强制重新选择批次货物。当默认批次货物数量小于等于0的时候，默认批次货物为有数量的，日期最早的货物批次。（默认批次A,数量100，日期06/10/2023. 批次B，数量300，日期09/10/2303.默认批次A数量小于等于0的时候。默认批次变成B）

**Reply by densen2014** [reply]
回复直接接在那个讨论话题下面,会比较清晰

![image](https://github.com/densen2014/docs/assets/8428709/8e2ccbed-0c35-4807-9a3c-f98be136490e)

**Comment by densen2014** [comment]
## 默认列表排序

1. 商品ID
2. 数量大于0的排在后面, 数量小于等于0的排在后面
3. 过期日,最早过期排在前面
4. 添加日期,最早过期排在前面
5. 库存小的排在后面
6. 批次号码短的在前面
7. 批次号码

```
                .OrderBy(a => a.ProductID)
                .OrderBy(a => a.Quantity<=0?true:false)
                .OrderBy(a => a.ExpiryDate)
                .OrderBy(a => a.AddDate)
                .OrderBy(a => a.Quantity)
                .OrderBy(a => a.BatchCode.Length)
                .OrderBy(a => a.BatchCode)
```

![image](https://github.com/densen2014/docs/assets/8428709/6efeffe9-2456-4f76-9bb1-265c6e9e9a10)

**Comment by densen2014** [comment]
## 选择批次逻辑

1. 选择批次的库存<0 提醒"库存不足, 当"库存不足也可提交"勾选,可添加到右边选择列表
2. 选择批次的库存小于(订单数量-选定批次总数量), 提醒"库存x不满足订单数量y,需要继续选择批次",添加到右边选择列表
3. 当选择批次的库存大于等于(订单数量-选定批次总数量),添加选择的批次的差额数量到右边选择列表,并走第四步骤自动提交
4. 提交按钮, 
   1. 选择批次为空提醒"请选择批次",
   2. 数量不等提醒"数量不足"
   3. 数量不等并"库存不足也可提交"勾选,提示"数量不足,但是选择了强制提交" 提交到销售单.
   4. 以上条件都满足则直接提交到销售单.
5. 销售窗口处理生成格式  $"{item.BatchCode} x {item.Quantity},..."


![image](https://github.com/densen2014/docs/assets/8428709/c72a01ab-a767-4b19-ba93-50fd8d11eaf3)
